### PR TITLE
merge arrays in createDocObject and changed object equality function

### LINKED ISSUE #7399: Documentation plugin will always write to full_documentation.json
**Describe the bug**
Every time strapi-plugin-documentation/config/functions/bootstrap.js runs, it will write to full_documentation.json even if no update is created.

**Steps to reproduce the behavior**

1. create a new project with users-permissions and documentation plugin installed 
2. start the server 
3. it should edit full_documentation.json but only changing the`x-generation-date` info

**Expected behavior**
If no update has been made, it should not write to the file

**Screenshots**

image
image

**Code snippets**
when bootstrap.js runs, it will almost always enter this condition
```js
// strapi-plugin-documentation/config/functions/bootstrap.js: line 101
if (!isMergedDocumentationExists || shouldUpdateFullDoc) {
    // Create the folder
    services.createDocumentationDirectory(documentationPath);
    // Write the file
    fs.writeFileSync(
      path.resolve(documentationPath, 'full_documentation.json'),
      JSON.stringify(fullDoc, null, 2),
      'utf8'
    );
}
```

as caused by checkIfPluginDocNeedsUpdate returning true when checking the "users-permissions" plugin, as it has 2 tags: UsersPermissions - User and UsersPermissions - Role
```js
// strapi-plugin-documentation/config/functions/bootstrap.js: line 38
const needToUpdatePluginDoc = services.checkIfPluginDocNeedsUpdate(plugin);
```

this happens because createDocObject() uses _.merge() which doesnt work if there is a property that contains an array, as it will override the array. Thus only one tag can appear in the object. To fix that the function should be changed to mergeWith that has a customizer for array values
```js
// strapi-plugin-documentation/services/Documentation.js
// from
createDocObject: array => {
    return array.reduce((acc, curr) => _.merge(acc, curr), {});
},

// to 
createDocObject: array => {
    function forArray(objValue, srcValue) {
      if (_.isArray(objValue)) {
        return objValue.concat(srcValue);
      }
    }
    return array.reduce((acc, curr) => _.mergeWith(acc, curr,forArray), {});
},
```

however this raises another error in the `areObjectsEquals()` function as it simply does `return JSON.stringify(obj1) === JSON.stringify(obj2);` to check for equality and will fail if the arrays don't have the exact same order

**System**
- Docker image: strapi/strapi:3.1.3
- Node.js version: v12
- NPM version:
- Strapi version: 3.1.3
- Database: MongoDB
- Operating system: Linux

### LINKED ISSUE #7509: x-generation-date gets updated on every server restart
**Describe the bug**

In the `./extensions/documentation/documentation/vx.x.x/full_documentation.json` file, the `x-generation-date` field under `info` gets updated every time the server is restarted.  This happens even when there has been no change to the documentation or any other code in the project's codebase. 

This causes problems because git tracks this file, so I am constantly committing this file for this `x-generation-date` field or I am forced to manually stage/unstage it from the rest of my changes.

**Steps to reproduce the behavior**

1. Have a strapi server running in development mode w/ the documentation plugin installed
2. restart the strapi server, again in development mode,  with no changes made to codebase
3. observe change in `x-generation date` 

**Expected behavior**
I would expect the `x-generation-date` to only update when a change or update to documentation occurs.

**Screenshots**
This is a git diff of the only line in the only file in the entire project that changes when the server gets restarted:

image

**System**

Docker-Compose Stack
- Node.js version: lts v12
- NPM version: 6
- Strapi version: strapi/strapi:3.1.1-node12
- Database: mongoDb
- Operating system: Ubuntu

**Additional context**
This occurs every time I restart the docker-compose stack.
- The stack runs my strapi project code in a mounted volume
- The strapi server is run in development mode

### PR DESCRIPTION
**Pull request by gh0stsh0t**
#### Description of what you did:
for: strapi-plugin-documentation
Changed the object equality method to use lodash's is equal to avoid unordered properties returning false
Changed createDocObject's merge to concat arrays instead of overwriting them
fix #7399
fix #7509

### CODE REVIEW

File: packages/strapi-plugin-documentation/services/Documentation.js [hunk 1]
```diff
@@ -20,7 +20,12 @@ const defaultSettingsKeys = Object.keys(defaultSettings);
 
 module.exports = {
   areObjectsEquals: (obj1, obj2) => {
-    return JSON.stringify(obj1) === JSON.stringify(obj2);
+    // stringify to remove nested empty objects
+    return _.isEqual(JSON.parse(JSON.stringify(obj1)), JSON.parse(JSON.stringify(obj2)))
```
Comment (alexandrebodin) [inline packages/strapi-plugin-documentation/services/Documentation.js:24]: Do you really need the JSON.parse of JSON.stringify here ?

Comment (gh0stsh0t) [reply packages/strapi-plugin-documentation/services/Documentation.js:24]: Hi, checking on this again, JSON.stringify works on its own apparently, but it isn't consistent, it still will fail if there is more than a single tag and isn't ordered in the same way. I had to create a custom comparator to check for arrays in the object
image 
image

**Comment by micahbule** [comment]
Any updates on this one? This is ruining auto-deploy scripts at the moment.

File: packages/strapi-plugin-documentation/services/Documentation.js [hunk 2]
```diff
@@ -20,7 +20,28 @@ const defaultSettingsKeys = Object.keys(defaultSettings);
 
 module.exports = {
   areObjectsEquals: (obj1, obj2) => {
-    return JSON.stringify(obj1) === JSON.stringify(obj2);
+    // stringify to remove nested empty objects
```
Comment (alexandrebodin) [inline packages/strapi-plugin-documentation/services/Documentation.js:23]: Hi starting from your code I made a simplified version that doesn't require a sort but runs a recursive compare instead. Can you refactor to match this logic ? 

```js
const cleanObject = obj => JSON.parse(JSON.stringify(obj));

const customComparator = (value1, value2) => {
  if (_.isArray(value1) && _.isArray(value2)) {
    if (value1.length !== value2.length) { return false };
    return value1.every(el1 => value2.findIndex(el2 => customIsEqual(el1, el2)) >= 0)
  }
}

const customIsEqual = (obj1, obj2) => _.isEqualWith(obj1, obj2, customComparator)

const areObjectsEquals = (obj1, obj2) => {
  return customIsEqual(cleanObject(obj1), cleanObject(obj2))
}
```

**Review by alexandrebodin — CHANGES_REQUESTED** [review]

File: packages/strapi-plugin-documentation/services/Documentation.js (hunk 2 above)
Comment (gh0stsh0t) [reply packages/strapi-plugin-documentation/services/Documentation.js:31]: Hi thanks for the review! Refactor should be pushed now

**Review by alexandrebodin — APPROVED** [review]
LGTM, Thank you !

**Comment by derrickmehaffy** [comment]
This pull request has been mentioned on **Strapi Community Forum**. There might be relevant details there:

https://forum.strapi.io/t/google-app-engine-readonly-vs-plugin-api-documentation-generation/468/7

**Comment by aquadesk** [comment]
I still get error in app engine
```
2021-08-30 05:16:14 default[20210829t214453]  [2021-08-30T05:16:14.616Z] error Error: EROFS: read-only file system, open '/workspace/extensions/documentation/documentation/1.0.0/full_documentation.json'
2021-08-30 05:16:14 default[20210829t214453]      at Object.openSync (fs.js:498:3)
2021-08-30 05:16:14 default[20210829t214453]      at Object.writeFileSync (fs.js:1524:35)
2021-08-30 05:16:14 default[20210829t214453]      at module.exports (/workspace/node_modules/strapi-plugin-documentation/config/functions/bootstrap.js:106:8)
2021-08-30 05:16:14 default[20210829t214453]      at processTicksAndRejections (internal/process/task_queues.js:95:5)
2021-08-30 05:16:14 default[20210829t214453]      at async Promise.all (index 2)
2021-08-30 05:16:14 default[20210829t214453]      at async Strapi.runLifecyclesFunctions (/workspace/node_modules/strapi/lib/Strapi.js:443:5)
2021-08-30 05:16:14 default[20210829t214453]      at async Strapi.load (/workspace/node_modules/strapi/lib/Strapi.js:379:5)
2021-08-30 05:16:14 default[20210829t214453]      at async Strapi.start (/workspace/node_modules/strapi/lib/Strapi.js:196:9)
2021-08-30 05:16:15 default[20210829t214453]  "GET /favicon.ico HTTP/1.1" 204
```

```
    "strapi": "3.6.8",
    "strapi-admin": "3.6.8",
    "strapi-connector-bookshelf": "3.6.8",
    "strapi-connector-firestore": "^3.0.0-alpha.40",
    "strapi-plugin-content-manager": "3.6.8",
    "strapi-plugin-content-type-builder": "3.6.8",
    "strapi-plugin-documentation": "3.6.8",
    "strapi-plugin-email": "3.6.8",
    "strapi-plugin-i18n": "3.6.8",
    "strapi-plugin-upload": "3.6.8",
    "strapi-plugin-users-permissions": "3.6.8",
    "strapi-provider-upload-google-cloud-storage": "3.6.5",
    "strapi-utils": "3.6.8"
```

**Comment by aquadesk** [comment]
nevermind, I generated the doc in localhost and push it to prod and it worked.
sorry!

**Comment by jkqubix** [comment]
For me this doesn't work. I have a Strapi installed, each time I deploy it to Google Cloud's App Engine Strapi tries to read/write into a documentation file. 

For example I have a custom api endpoint under docs. Strapi apparently tries to generate documentation for it:
image

This happens also with all Strapi generated content-types I have (removing the docs api endpoint doesn't help). 

I have also observed that every time I start the server on localhost "x-generation-date" gets regenerated. That could also cause a problem (and I have seen this in a different issue, apparently marked as fixed).

I use Strapi and Documentation with version 4.1.10. 

Is there a way how to make the documentation generation manual or turn it off (e. g. to have it only when I launch Strapi in develop mode) and prevent documentation generation on production?

**Pull request merged**